gapis/perfetto/android/trace.go

```diff
@@ -132,36 +132,15 @@ func Start(ctx context.Context, d adb.Device, a *android.ActivityAction, opts *s
 			"activity": a.Activity,
 		}.Bind(ctx)
 
-		packages := []string{}
-		driver, err := d.GraphicsDriver(ctx)
-		if err != nil {
-			// If there's an error, keep going to attempt to use GPU profiling
-			// libraries in system image.
-			log.W(ctx, "Failed to query developer driver: %v, assuming no developer driver found.", err)
-		}
-		if driver.Package != "" {
-			log.I(ctx, "Using GPU profiling libraries from developer driver package: %v.", driver.Package)
-
-			// Setup the application to use the prerelease driver.
-			nextCleanup, err := adb.SetupPrereleaseDriver(ctx, d, a.Package)
-			cleanup = cleanup.Then(nextCleanup)
-			if err != nil {
-				return nil, cleanup.Invoke(ctx), err
-			}
-			packages = append(packages, driver.Package)
-		} else {
-			log.I(ctx, "No developer driver found, attempting to use GPU profiling libraries in system image.")
-			if supported, err := d.HasGpuProfilingSupportInSystemImage(ctx); err != nil || !supported {
-				return nil, cleanup.Invoke(ctx), log.Err(ctx, err, "No developer driver found, and no GPU profiling support found in system image.")
-			}
-			if vulkanLayerApk, err := d.GetGpuProfilingLayerPackageName(ctx); err == nil && vulkanLayerApk != "" {
-				packages = append(packages, vulkanLayerApk)
-			}
+		supported, packageName, nextCleanup, err := d.PrepareGpuProfiling(ctx, a.Package)
+		cleanup = cleanup.Then(nextCleanup)
+		if err != nil || !supported {
+			return nil, cleanup.Invoke(ctx), log.Err(ctx, err, "GPU profiling is not supported")
 		}
-		hasRenderStages := hasDataSourceEnabled(opts.PerfettoConfig, gpuRenderStagesDataSourceName)
 
 		// Setup the profiling layers.
-		nextCleanup, err := setupProfileLayers(ctx, d, a.Package.Name, hasRenderStages, abi, packages, layers)
+		hasRenderStages := hasDataSourceEnabled(opts.PerfettoConfig, gpuRenderStagesDataSourceName)
+		nextCleanup, err = setupProfileLayers(ctx, d, a.Package.Name, hasRenderStages, abi, []string{packageName}, layers)
```

Comment:
Same thing here, and here I think it actually would be a problem, since in `setupProfileLayers` we append another string to the list, and thus, if `packageName` is empty, it would try to set the property to `:<gapid.package>`, which is more likely going to fail than succeed.

Reply:
Thanks for this, I noticed a similar issue existing before my refactor in `SetupProfileLayersSource()`, where we don't check if `driver.Package` is an empty string or not: https://github.com/google/agi/blob/master/gapis/perfetto/android/trace.go#L76-L80

Rather than checking in various places that we don't add an empty string in the layer/package lists, I decided to cleanup these two lists directly in `android.SetupLayers()` by removing empty names.

Let me know if that works for you.